lisa/operating_system.py

```diff
@@ -150,36 +221,110 @@ def type_name(cls) -> str:
+            if not os_release_info:
+                continue
+            if os_release_info.group("name") == "NAME":
+                os_version.vendor = os_release_info.group("value")
+            elif os_release_info.group("name") == "VERSION_ID":
+                os_version.release = os_release_info.group("value")
+            elif os_release_info.group("name") == "VERSION":
+                os_version.codename = get_matched_str(
+                    os_release_info.group("value"),
+                    self.__distro_codename_pattern,
+                )
+
+        if os_version.vendor == "":
+            raise LisaException("OS version information not found")
+
+        return os_version
+
+    def _install_package_from_url(
+        self,
+        package: str,
+        signed: bool = True,
+    ) -> None:
+        """
+        Used if the package to be installed needs to be downloaded from a url first.
+        """
+        # when package is URL, download the package first at the working path.
+        wget_tool = self._node.tools[Wget]
+        pkg = wget_tool.get(package, str(self._node.working_path))
+        self.install_packages(pkg, signed)
+
     def install_packages(
-        self, packages: Union[str, Tool, Type[Tool], List[Union[str, Tool, Type[Tool]]]]
+        self,
+        packages: Union[str, Tool, Type[Tool], List[Union[str, Tool, Type[Tool]]]],
+        signed: bool = True,
     ) -> None:
         package_names: List[str] = []
         if not isinstance(packages, list):
             packages = [packages]
 
         assert isinstance(packages, list), f"actual:{type(packages)}"
         for item in packages:
-            if isinstance(item, str):
-                package_names.append(item)
-            elif isinstance(item, Tool):
-                package_names.append(item.package_name)
-            else:
-                assert isinstance(item, type), f"actual:{type(item)}"
-                # Create a temp object, it doesn't trigger install.
-                # So they can be installed together.
-                tool = item.create(self._node)
-                package_names.append(tool.package_name)
+            package_names.append(self.__resolve_package_name(item))
         if self._first_time_installation:
             self._first_time_installation = False
             self._initialize_package_installation()
-        self._install_packages(package_names)
+
+        self._install_packages(package_names, signed)
+
+    def package_exists(
+        self, package: Union[str, Tool, Type[Tool]], signed: bool = True
+    ) -> bool:
+        """
+        Query if a package/tool is installed on the node.
+        Return Value - bool
+        """
+        package_name = self.__resolve_package_name(package)
+        return self._package_exists(package_name)
+
+    def update_packages(
+        self, packages: Union[str, Tool, Type[Tool], List[Union[str, Tool, Type[Tool]]]]
+    ) -> None:
+        raise NotImplementedError
```

Comment:
Any use case for this? What's the role of a test suite in updating stuff?

Reply:
Looking good otherwise :)

Reply:
It might be used internally say - in case the some package/tool version queried doesn't match the required version then it would need to be updated first.
But yeah it most likely should not be taking in a list.